### PR TITLE
add some helpers to deal with streams of bytes

### PR DESCRIPTION
**Pull request by jap**
This PR adds two little helpers that make it easier to parse streams of bytes.

If this is something that stands a chance of being merged, I can also spend some time on adding docs.

(It also contains an unrelated update to the pre-commit config to make the flake8 plugin work again.)

### CODE REVIEW

**Comment by spookylukey** [comment]
This looks like a useful addition. Some questions (to which I don't know the answer, just thinking out loud):

- should it be `bconcat` or `concatb` etc? Is there prior art in stdlib or somewhere?
  -  An argument for `concatb`- in terms of discoverability in IDEs that give completion, it's more obvious that you have two variants if you see `concat` and `concatb` next to each other. Similarly in docs, if they are next to each other you could avoid repeating everything. Maybe there are other arguments though.
- are there other utilities that need the same treatment?

File: src/parsy/__init__.py [hunk 1]
```diff
@@ -538,6 +545,27 @@ def string_parser(stream, index):
     return string_parser
 
 
+def bstring(expected_bytes: bytes, transform: Callable[[bytes], bytes] = noop) -> Parser:
```
Comment (wbolster) [inline src/parsy/__init__.py:548]: why have this as a separate function? this code is a direct copy of `parsy.string()`, which already works just fine on binary strings (`bytes`):

``` python
>>> p = parsy.string(b'abc')
>>> p.parse_partial(b'abcdef')
(b'abc', b'def')
```

… except for its type annotation which contains `str` (a few times). changing those annotations to [`typing.AnyStr`](https://docs.python.org/3/library/typing.html#typing.AnyStr) should be enough.

**Comment by jap** [comment]
@wbolster 's comment makes sense! So I've just amended this commit to remove `bstring` again, but kept the tests and modified the type annotation of `string` to also accept bytes. (The function name is a bit confusing now but I can live with that.)

In a similar vein, I changed `.concat` to work on both lists of str and bytes, but unfortunately that was a bit more involved, and required writing out the bind/map operation because the mapped function needs access to the type of the input stream :(
The tests from the `bconcat` version still work though.

File: src/parsy/__init__.py [hunk 2]
```diff
@@ -166,10 +166,23 @@ def combine_dict(self, combine_fn: Callable) -> Parser:
...
-        produced values.
+        Returns a parser that concatenates together the previously produced values.
+
+        This parser will join the values using the type of the input stream, so
+        when feeding bytes to the parser, the items to be joined must also be bytes.
         """
-        return self.map("".join)
+
+        @Parser
+        def parser(stream: bytes | str, index: int) -> Result:
```
Comment (wbolster) [inline src/parsy/__init__.py:176]: this could be `typing.AnyStr` btw. (that would retain the type if `Result` would become a generic type)

**Comment by wbolster** [comment]
@spookylukey any chance you can have another look? in its current form this PR does not change any API, and just makes more stuff transparently work with both unicode strings (`str`) and binary strings (`bytes`)

**Comment by spookylukey** [comment]
This looks good, thanks!

**Comment by spookylukey** [comment]
I'm afraid I had to revert this.

Somehow in the middle of the night I realised the approach in this PR can't work. It makes the assumption that the type of the current parser's return values will be the type of the input stream, but there is no reason why this should be the case. Looking again, I see that this is documented in the new docstring, but it's a backwards incompatible breaking change.

I've added a test which demonstrates this - on reverting the change in this PR, the test passes, but not otherwise. We can't assume that no-one is writing code like in the test.

I can't see there is any way to get this to work. You can't test the type of the current return value either, as you might have zero items - an empty list of `str` is the same as an empty list of `byte`.

A backwards compatible alternative would be to do:

```python
    def concat(self, joiner="") -> Parser:
        return self.map(joiner.join)
```

Would that work for you? Otherwise it's back to thinking about `concatb` etc.

**Comment by wbolster** [comment]
hmmm 🤔

adding an argument to `.concat()` to indicate how the joining should happen takes away most of the ‘it just works’ convenience, so i don't think it's really worth it in that case, as one could as well use

``` python
parser.map(b"".join)
```

… instead, which is more explicit, more flexible, and also shorter than the suggested

``` python
parser.concat(self, joiner=b"") 
```

**Comment by spookylukey** [comment]
So we could go with `concatb()`, or we could also put this in the docstring of `concat()`:

> Synonym for `.map("".join)`

Which would give people a clue about how to easily implement `concatb` themselves

**Pull request merged**